### DISCUSSION TITLE
Dependency injection does not work as expected (?)

### DISCUSSION THREAD

**Discussion by edward-b-1** [fastapi/fastapi#11894] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
########################
# fastapi_webserver.py #
########################

from fastapi import FastAPI
from fastapi import Request
from fastapi import Depends

from httpx import AsyncClient

from contextlib import asynccontextmanager

from implementation import Implementation
from datetime_real import DatetimeReal
from get_implementation import get_implementation
from get_datetime_strategy import get_datetime_strategy


@asynccontextmanager
async def lifespan(app: FastAPI):
    async with AsyncClient(app=app) as client:
        implementation = Implementation()
        datetime_strategy = DatetimeReal()
        yield {
            'implementation': implementation,
            'datetime_strategy': datetime_strategy,
        }
        implementation.shutdown() # MUST be called!

app = FastAPI(lifespan=lifespan)

@app.get('/api/get_value')
async def api_get_value(
    request: Request,
    implementation:Implementation = Depends(get_implementation),
    datetime_strategy:DatetimeReal = Depends(get_datetime_strategy),
):
    print(request.client.host)
    print(f'{datetime_strategy.now()}') # 2024-07-26 09:00:00+00:00 (expected value)

    now, value = implementation.get_value()
    return {
        'now': now, # 2024-07-26T10:03:35.718263+00:00 (wrong value)
        'value': value,
    }


###################
# example_test.py #
###################

import pytest

from fastapi.testclient import TestClient

from httpx import AsyncClient

from fastapi_webserver import app

from implementation_fake import ImplementationFake
from get_implementation import get_implementation

from datetime_fake import DatetimeFake
from get_datetime_strategy import get_datetime_strategy

from datetime import datetime
from datetime import timezone


implementation_fake = ImplementationFake()

async def get_implementation_fake():
    return implementation_fake

app.dependency_overrides[get_implementation] = get_implementation_fake


datetime_fake = DatetimeFake()

async def get_datetime_fake():
    return datetime_fake

app.dependency_overrides[get_datetime_strategy] = get_datetime_fake


def test_get_value():
    fixed_datetime_str = (
        str(
            datetime(
                year=2024, month=7, day=26,
                hour=9, minute=0, second=0,
                tzinfo=timezone.utc,
            )
        )
    )

    with TestClient(app) as client:
        response = client.get('/api/get_value')
        assert response.status_code == 200
        assert response.json() == {
            'now': fixed_datetime_str, # wrong value !
            'value': 42,
        }


#####################
# implementation.py #
#####################

from datetime import datetime

from datetime_real import DatetimeReal

class Implementation():

    def __init__(self) -> None:
        self._datetime_strategy = DatetimeReal()
        self.current_value = 0
        self._initialized = True

    def shutdown(self) -> None:
        '''
        A function which must be called to cleanup resources before exit
        '''
        self._initialized = False

    def increment(self) -> datetime:
        '''
        In reality this would do something like read/write to a file, db etc
        '''
        self.current_value += 1
        _now = self._datetime_strategy.now()
        return _now

    def get_value(self) -> tuple[datetime, int]:
        _now = self._datetime_strategy.now()
        return _now, self.current_value


##########################
# implementation_fake.py #
##########################

from datetime import datetime

from datetime_real import DatetimeReal


class ImplementationFake():

    def __init__(self) -> None:
        self._datetime_strategy = DatetimeReal()

    def shutdown(self) -> None:
        '''
        A function which must be called to cleanup resources before exit
        (at least if the implementation being used is the real implementation
        - the fake implementation probably wouldn't do anything here)
        '''
        pass

    def increment(self) -> datetime:
        _now = self._datetime_strategy.now()
        return _now

    def get_value(self) -> tuple[datetime, int]:
        _now = self._datetime_strategy.now()
        return _now, 42


####################
# datetime_real.py #
####################

from datetime import datetime
from datetime import timezone

class DatetimeReal():
    def now(self) -> datetime:
        return (
            datetime.now(timezone.utc)
        )


####################
# datetime_fake.py #
####################

from datetime import datetime
from datetime import timezone

class DatetimeFake():
    def now(self) -> datetime:
        return (
            datetime(
                year=2024, month=7, day=26,
                hour=9, minute=0, second=0,
                tzinfo=timezone.utc,
            )
        )


#########################
# get_implementation.py #
#########################

from fastapi import Request

from implementation import Implementation

async def get_implementation(
    request: Request,
):
    implementation: Implementation = request.state.implementation
    return implementation


############################
# get_datetime_strategy.py #
############################

from fastapi import Request

from datetime_real import DatetimeReal

async def get_datetime_strategy(
    request: Request,
):
    datetime_strategy: DatetimeReal = request.state.datetime_strategy
    return datetime_strategy
```

### Description

I have written an example code to figure out how dependency injection is supposed to work. In this example code, I have two dependencies.

- There is an `Implementation` which is intended to represent a concept of some "system" which FastAPI provides an interface to
- A datetime dependency. This is really supposed to be a "global" dependency in the same way that `datetime.now()` is globally accessible (provided you import it)

To be completely honest, I'm not totally sure if my idea for how to write a substitutable dependency for `datetime.now()` is really the right approach. I have built two classes. However, rather than calling a function to get the datetime, clients must now create a `DatetimeReal` or `DatetimeFake` object, and then call the method `.now()`. This function takes a `self`, so it's a bit weird.

Regardless of this, the datetime dependency does not appear to be working correctly. Rather than returning a fake, fixed, datetime value, the real datetime is being returned from within the example test code.

I am unsure as to why this is happening. I suspect that I am using dependency in the wrong way or otherwise misunderstand how it should be used. Possibly having nested dependencies does not work? (The datetime strategy is a dependency of the `Implementation` object. The FastAPI code really only depends on `Implementation` directly, and the datetime strategy _independently_.)

It would be great to get some feedback on whether this design is the right approach and if not what I should be doing instead. Basically - I don't understand why this doesn't work as expected.

### Operating System

Linux

### FastAPI Version

0.111.1

### Pydantic Version

2.8.2

### Python Version

3.11.2

**Comment by YuriiMotov** [comment]
Please, provide [MRE](https://en.wikipedia.org/wiki/Minimal_reproducible_example).
I see lots of questions from you at StackOverflow, and they are always quite verbose and contain more code than it's needed to demonstrate the problem.

**Reply by edward-b-1** [reply]
This is an MRE - it can't be reduced any further. The "real" project I am working on is hundreds of files, 10k + lines

**Reply by YuriiMotov** [reply]
>  it can't be reduced any further

It can be reduced. You either don't want to do it or don't know how to do it. But it can be easily reduced.

You have an assumption about why your code doesn't work. Try to start from scratch and write the most simple code that demonstrates your problem.
Your mistake is actually quite simple. I think you will find the answer while you are trying to write this MRE

**Reply by edward-b-1** [reply]
Ok, I can see this isn't going to be as productive as I had hoped. I don't wish to engage with a confrontation regarding this. I'm open to suggestions as to how I could reduce the code further. At the moment I don't see how any of it could be removed without it not fully being representative of the other, much larger, code which I am working with. I have already gone to the effort of abstracting away the unimportant details to write this MWE.

The code provided here, including comments, blank lines and imports is a mere 230 lines. That is spread across 8 files. Hardly a large example code.

> Try to start from scratch and write the most simple code that demonstrates your problem.

This is exactly what I have already done to produce this.

> Your mistake is actually quite simple.

It would be great if you could tell me what it is. Presumably if it is simple it would be simple to explain?

**Reply by YuriiMotov** [reply]
> It would be great if you could tell me what it is. 

I have already answered a couple of your questions on StackOverflow recently. And you haven't even said "Thank you".

Looks like you treat answers on StackOverflow and here the same way as the answers of ChatGPT. This is incorrect.

**Reply by edward-b-1** [reply]
Sorry if I have personally offended you. It was not intentional.

> And you haven't even said "Thank you".

This is not true, I can find several comments where I thanked you for your help.

I didn't manage to get the "full" answer from previous questions, so I decided to put a full MWE here. This platform is probably more conducive to obtaining productive information. Stack Overflow has its problems.

**Reply by YuriiMotov** [reply]
As for reducing the code example:

1) What if you remove `lifespan` and use global object instead? Does it still demonstrates your problem?

2) What if you place your code in one file and remove unnecessary import lines? It will be less lines of code and this way it's easier for people to copy and run your code.

I have more ideas how to simplify your code, but I think those two can show you the direction to move.

**Reply by edward-b-1** [reply]
> 1. What if you remove `lifespan` and use global object instead? Does it still demonstrates your problem?

I can't do this. The "state" has to be managed by a `lifespan` such that `.shutdown()` is called. If it were a global object, `.shutdown()` would not be called when a SIGINT occurs.

**Reply by YuriiMotov** [reply]
> I can't do this. The "state" has to be managed by a `lifespan` such that `.shutdown()` is called. If it were a global object, `.shutdown()` would not be called when a SIGINT occurs.

Do you think it's important for MRE?

**Reply by edward-b-1** [reply]
> Do you think it's important for MRE?

Yes, it is critical. There is absolutly no point in asking a question without this, because when this constraint is introduced it changes the context.

**Reply by edward-b-1** [reply]
```
from fastapi import FastAPI
from fastapi import Request
from fastapi import Depends
from fastapi.testclient import TestClient

from httpx import AsyncClient

from contextlib import asynccontextmanager

from datetime import datetime
from datetime import timezone


class DatetimeReal():
    def now(self) -> datetime:
        return (
            datetime.now(timezone.utc)
        )


class DatetimeFake():
    def now(self) -> datetime:
        return (
            datetime(
                year=2024, month=7, day=26,
                hour=9, minute=0, second=0,
                tzinfo=timezone.utc,
            )
        )


class Implementation():

    def __init__(self) -> None:
        self._datetime_strategy = DatetimeReal()
        self.current_value = 0
        self._initialized = True

    def shutdown(self) -> None:
        '''
        A function which must be called to cleanup resources before exit
        '''
        self._initialized = False

    def increment(self) -> datetime:
        '''
        In reality this would do something like read/write to a file, db etc
        '''
        self.current_value += 1
        _now = self._datetime_strategy.now()
        return _now

    def get_value(self) -> tuple[datetime, int]:
        _now = self._datetime_strategy.now()
        return _now, self.current_value


class ImplementationFake():

    def __init__(self) -> None:
        self._datetime_strategy = DatetimeReal()

    def shutdown(self) -> None:
        '''
        A function which must be called to cleanup resources before exit
        (at least if the implementation being used is the real implementation
        - the fake implementation probably wouldn't do anything here)
        '''
        pass

    def increment(self) -> datetime:
        _now = self._datetime_strategy.now()
        return _now

    def get_value(self) -> tuple[datetime, int]:
        _now = self._datetime_strategy.now()
        return _now, 42


async def get_implementation(
    request: Request,
):
    implementation: Implementation = request.state.implementation
    return implementation


async def get_datetime_strategy(
    request: Request,
):
    datetime_strategy: DatetimeReal = request.state.datetime_strategy
    return datetime_strategy


@asynccontextmanager
async def lifespan(app: FastAPI):
    async with AsyncClient(app=app) as client:
        implementation = Implementation()
        datetime_strategy = DatetimeReal()
        yield {
            'implementation': implementation,
            'datetime_strategy': datetime_strategy,
        }
        implementation.shutdown() # MUST be called!


app = FastAPI(lifespan=lifespan)


@app.get('/api/get_value')
async def api_get_value(
    request: Request,
    implementation:Implementation = Depends(get_implementation),
    datetime_strategy:DatetimeReal = Depends(get_datetime_strategy),
):
    print(request.client.host)
    print(f'{datetime_strategy.now()}') # 2024-07-26 09:00:00+00:00 (expected value)

    now, value = implementation.get_value()
    return {
        'now': now, # 2024-07-26T10:03:35.718263+00:00 (wrong value)
        'value': value,
    }


implementation_fake = ImplementationFake()

async def get_implementation_fake():
    return implementation_fake

app.dependency_overrides[get_implementation] = get_implementation_fake


datetime_fake = DatetimeFake()

async def get_datetime_fake():
    return datetime_fake

app.dependency_overrides[get_datetime_strategy] = get_datetime_fake


def test_get_value():
    fixed_datetime_str = (
        str(
            datetime(
                year=2024, month=7, day=26,
                hour=9, minute=0, second=0,
                tzinfo=timezone.utc,
            )
        )
    )

    with TestClient(app) as client:
        response = client.get('/api/get_value')
        assert response.status_code == 200
        assert response.json() == {
            'now': fixed_datetime_str, # wrong value !
            'value': 42,
        }
```

**Reply by edward-b-1** [reply]
I combined all the files together into a single test file. The results are exactly the same. The test still does not pass, the same issue is present as before.

When I asked this question, I said that I suspect I am not understand how the FastAPI framework is intended to be used. I still suspect this is the case.

Hopefully I did not make any mistakes when combining all these files into a single file. I have included the full code that I am working with above.

**Reply by YuriiMotov** [reply]
> Yes, it is critical. There is absolutly no point in asking a question without this, because when this constraint is introduced it changes the context.

Does it really change the situation dramatically if you change this part of code:

```
class Implementation():

    def __init__(self) -> None:
        self._datetime_strategy = DatetimeReal()
        self.current_value = 0
        self._initialized = True

    def shutdown(self) -> None:
        '''
        A function which must be called to cleanup resources before exit
        '''
        self._initialized = False

    def increment(self) -> datetime:
        '''
        In reality this would do something like read/write to a file, db etc
        '''
        self.current_value += 1
        _now = self._datetime_strategy.now()
        return _now

    def get_value(self) -> tuple[datetime, int]:
        _now = self._datetime_strategy.now()
        return _now, self.current_value


class ImplementationFake():

    def __init__(self) -> None:
        self._datetime_strategy = DatetimeReal()

    def shutdown(self) -> None:
        '''
        A function which must be called to cleanup resources before exit
        (at least if the implementation being used is the real implementation
        - the fake implementation probably wouldn't do anything here)
        '''
        pass

    def increment(self) -> datetime:
        _now = self._datetime_strategy.now()
        return _now

    def get_value(self) -> tuple[datetime, int]:
        _now = self._datetime_strategy.now()
        return _now, 42


async def get_implementation(
    request: Request,
):
    implementation: Implementation = request.state.implementation
    return implementation


async def get_datetime_strategy(
    request: Request,
):
    datetime_strategy: DatetimeReal = request.state.datetime_strategy
    return datetime_strategy


@asynccontextmanager
async def lifespan(app: FastAPI):
    async with AsyncClient(app=app) as client:
        implementation = Implementation()
        datetime_strategy = DatetimeReal()
        yield {
            'implementation': implementation,
            'datetime_strategy': datetime_strategy,
        }
        implementation.shutdown() # MUST be called!


app = FastAPI(lifespan=lifespan)
```
to this code:
```
class Implementation():

    def __init__(self) -> None:
        self._datetime_strategy = DatetimeReal()
        self.current_value = 0
        self._initialized = True

    def get_value(self) -> tuple[datetime, int]:
        _now = self._datetime_strategy.now()
        return _now, self.current_value


class ImplementationFake():

    def __init__(self) -> None:
        self._datetime_strategy = DatetimeReal()

    def get_value(self) -> tuple[datetime, int]:
        _now = self._datetime_strategy.now()
        return _now, 42


_implementation = Implementation()
_datetime_strategy = DatetimeReal()


async def get_implementation():
    return _implementation


async def get_datetime_strategy():
    return _datetime_strategy


app = FastAPI()
```

There are a lot of similar things you can do to reduce the size of your MRE.

When you remove unnecessary code, it will be easier for others to help you. And it will be easier for you to find the problem as well.

**Reply by YuriiMotov** [reply]
More things you can do:

3) Simplify `DatetimeReal` and `DatetimeFake` by returning `1` and `2` instead of dates

4) Remove unnecessary prints and dependency from `api_get_value`

5) Move all code that is related to tests (fake classes, dependency override) to the bottom of the file. So that it would be easier to understand the logic you want to implement

Do these things and edit your initial post (past new MRE there) and I'm sure you will have more people who wants to help you

**Reply by edward-b-1** [reply]
> 4\. Remove unnecessary prints and dependency from `api_get_value`

Those are not uneccessary. They are there to demonstrate to the person who runs the example code that it has the expected behaviour at this point, but not elsewhere in the program.

> 3\. Simplify DatetimeReal and DatetimeFake by returning 1 and 2 instead of dates

This is not a sensible suggestion. Changing the type does nothing to reduce the number of lines of code in the example.

> 5\. Move all code that is related to tests (fake classes, dependency override) to the bottom of the file. So that it would be easier to understand the logic you want to implement

Again - not sensible. Doesn't change the number of lines of code. The code won't even run as a single file unless things are declared in the right order. If working with single files, the order of which file comes first is irrelevant.

> Do the these things and edit your initial post (past new MRE there) and I'm sure you will have more people who wants to help you

I literally did the things you asked, and combined everything into a single file for you. You're still not happy, and behaving quite unreasonably, actually.

A few comments earlier in this thread you essentially said that you knew the answer to what was wrong, but are refusing to tell me.

It's very strange behavior. I believe you were asking for help to find a job recently. It might be worth considering that employers will use social media sites to do background checks on candidates. That includes sites like Github. (At least this has been my experience.) If you behave in an obstructive manner, or just generally behave in a way which causes frustration, that probably isn't going to reflect very well.

I've been polite and respectful towards you, and I don't believe I've done anything to cause you to be offended in any way. If you're unwilling to help answer my question then I would respectfully ask you to simply not answer it, and to ignore it. This is preferable to a continual back and fourth where you repeatedly reply in a combative manner as if you are trying to find just one thing to which you can say "I'm right and he's wrong". It's just pointless. I already told you I don't know how to use FastAPI. There's no need to try and prove a point here.

**Reply by YuriiMotov** [reply]
I'm not trying to show that I know FastAPI and you don't.
As I said before, I don't like your way of asking questions (1. lots of long questions; 2. not saying "thank you" (don't read it literally, it's not only about phrase, you have the ability to press "accept answer" and upvote answer)).
And I want to help you learn how to ask questions in a way that can increase you chances to get the answer.

I just don't want to simply give you the answer (partially it's because I'm a little angry on you). I want to lead you to that answer. And I think you will find the answer if you simplify your MRE

**Reply by YuriiMotov** [reply]
> This is not a sensible suggestion. Changing the type does nothing to reduce the number of lines of code in the example.

Let's compare:

```
class DatetimeReal():
    def now(self) -> datetime:
        return (
            datetime.now(timezone.utc)
        )


class DatetimeFake():
    def now(self) -> datetime:
        return (
            datetime(
                year=2024, month=7, day=26,
                hour=9, minute=0, second=0,
                tzinfo=timezone.utc,
            )
        )

@app.get('/api/get_value')
async def api_get_value(
    request: Request,
    implementation:Implementation = Depends(get_implementation),
    datetime_strategy:DatetimeReal = Depends(get_datetime_strategy),
):
    print(request.client.host)
    print(f'{datetime_strategy.now()}') # 2024-07-26 09:00:00+00:00 (expected value)

    now, value = implementation.get_value()
    return {
        'now': now, # 2024-07-26T10:03:35.718263+00:00 (wrong value)
        'value': value,
    }

def test_get_value():
    fixed_datetime_str = (
        str(
            datetime(
                year=2024, month=7, day=26,
                hour=9, minute=0, second=0,
                tzinfo=timezone.utc,
            )
        )
    )

    with TestClient(app) as client:
        response = client.get('/api/get_value')
        assert response.status_code == 200
        assert response.json() == {
            'now': fixed_datetime_str, # wrong value !
            'value': 42,
        }

```
VS
```
class DatetimeReal():
    def now(self) -> int:
        return 1


class DatetimeFake():
    def now(self) -> int:
        return 2

@app.get('/api/get_value')
async def api_get_value(
    request: Request,
    implementation:Implementation = Depends(get_implementation),
    datetime_strategy:DatetimeReal = Depends(get_datetime_strategy),
):
    print(request.client.host)
    print(f'{datetime_strategy.now()}') # 2 (expected value)

    now, value = implementation.get_value()
    return {
        'now': now, # 1 (wrong value)
        'value': value,
    }

def test_get_value():

    with TestClient(app) as client:
        response = client.get('/api/get_value')
        assert response.status_code == 200
        assert response.json() == {
            'now': 2, # wrong value !
            'value': 42,
        }
```

Result: -17 lines of code

**Reply by YuriiMotov** [reply]
> Those are not uneccessary. They are there to demonstrate to the person who runs the example code that it has the expected behaviour at this point, but not elsewhere in the program.

Actually, it doesn't demonstrate anything, but may just mislead you and others.

> Again - not sensible. Doesn't change the number of lines of code. The code won't even run as a single file unless things are declared in the right order. If working with single files, the order of which file comes first is irrelevant.

That doesn't change the number of lines, but that can improve the readability of code.
And, it's not impossible to move things as I mentioned above. Just try and ask if you have any troubles with it.

**Reply by edward-b-1** [reply]
For anyone who might be reading this in future (although this seems somewhat unlikely considering the thread was derailed by the vast amount of spam) **it might be the case that this dependency injection framework does not work in the way you expected it to**.

It's actually very limited. I'm not sure I can explain it clearly as there are a lot of subtleties here, but I will try my best.

The most important point is that the dependency overrides will only apply - let's say **"directly"** - meaning that the dependency overrides are only applied from the top most level of a FastAPI path definition.

To explain in more detail, here is an example path definition:

```
@app.post('/api/example_endpoint')
async def example_endpoint_definition(
    fastapi_order_insert_message: FastAPI_OrderInsertMessage,    # some type derived from pydantic.BaseModel
    request: Request,
    response: Response,
    webserver: Webserver = Depends(get_webserver_instance),    # first dependency
    now: Callable = Depends(get_now_function),    # second dependency
):
    # dependency at top level - works as expected
    timestamp = now()
    ip = request.client.host

    log.info(f'POST /api/example_endpoint ({ip}, {timestamp})')

    # dependency at top level - works as expected
    return webserver.send_order(fastapi_order_insert_message)
```

In this example, we are at the top level of the FastAPI code. We are still within the function which defines a path. Everything works as expected here.

What does not work as I had anticipated are any dependencies which are present further down the call stack.

For example, the **implementation of** `webserver.send_order(fastapi_order_insert_message)` **depends on the function** `now` **to get the current datetime**.

The implementation of this looks something like this:

```
def send_order(
    self,
    fastapi_order_insert_message: FastAPI_OrderInsertMessage,
) -> FastAPI_ReturnStatusWithTradesAndOrderId:

    order_insert_message = convert_fastapi_message_to_internal_message(
        fastapi_order_insert_message,
        timestamp=now(),
    )
```

This code lives in a different file.

As you can see, there is a dependency here on the function `now()`. **This dependency does not work as expected, if the top level code above is using a fake implementation, this block of code will still be using the real implementation.**

I don't know if this is wrong or not. It could be a bug in FastAPI, but I suspect that the dependency system of FastAPI is actually more limited than I initially had thought.

To provide a bit of further detail, the following block of code shows how `now` is resolved. This block of code lives in the same file as the above block.

```
def now() -> datetime:
    now_impl = get_now_function()
    return now_impl()
```

`get_now_function` is defined elsewhere. This is the dependency which I was expecting would be over-ridden, but it isn't.

To summarize:

- It is not possible to override a "global" dependency (like a call to `datetime.now()`)
- All objects which are dependencies which should be substituted for a different implementation within test code must be defined using `Depends` at the top level of FastAPI code (path definitions)
- Those objects must be passed around to all other code which needs to have the same dependency resolution behaviour
- This is clearly no good if one or more of those dependencies are global objects (like `datetime.now()`, or a logger) because those objects should not be passed around as function arguments
- So, it seems that if you have such a global dependency, you will have to roll your own dependency resolution code. You will have to figure out how to make it work yourself. FastAPI is not going to help you here

To explain the second bullet point in more detail:

- You likely have some objects in your FastAPI path definition code which need to be substituted for an alternative implementation when running tests
- These objects are "dependencies"
- Assuming they are **not global dependencies**, define them using `Depends` in the path
- If there are multiple objects, say for example `A` and `B`, then if some code inside `A` depends on `B`, you must pass `B` as an argument to that code

I hope that's a bit clearer. It might not be.

**Comment by YuriiMotov** [comment]
Closing this. Hope it has been solved
